### PR TITLE
[WIP] Improve tolerance for inertia triangle inequality check.

### PR DESCRIPTION
**Pull request by mitiguy**
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551)

### CODE REVIEW

**Review by mitiguy — COMMENTED** [review]
PR #19551 helps resolves issue #19544.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-:-NXIeP1tD2NfGyScEBp3:b3u669f)* status: needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @mitiguy)

**Review by mitiguy — COMMENTED** [review]
Feature review @sherm1 
Notes: 
1. The comments in spatial_inertia_test.cc may be repetitious.  Let me know.
2. Change Python bindings due to new optional parameter to RotationalInertia::CouldBePhysicallyValid().
 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-:-NXIgqQa8w70ysswOuAi:b-amdf1x)* status: needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

**Review by mitiguy — COMMENTED** [review]
@sherm1  Reduced repetitious comments in spatial_inertia_test.cc and added Python binding for new optional parameter.  

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-:-NXMltcJ4-mwoFqylI90:b-6ctvwt)* status: needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

**Review by sherm1 — COMMENTED** [review]
Feature review +@sherm1

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-:-NXNG2vCAMQX9H38dPZv:bet2hx0)* status: LGTM missing from assignee sherm1(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

**Review by sherm1 — COMMENTED** [review]
Looks great, thanks! One discussion point -- I think we should consider an absolute cutoff for the error test so that we don't accept something that has bad downstream consequences. PTAL at my comment below.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-:-NXQvdco-ZOwLJGr8Q51:b-u55jch)* status: 2 unresolved discussions, LGTM missing from assignee sherm1(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

___
*[`multibody/tree/rotational_inertia.h` line 591 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQsH870Um-DKl4XNZ8:-NXQsH870Um-DKl4XNZ9:b7jhuc7) ([raw file](https://github.com/RobotLocomotion/drake/blob/b84ac3b9d5ea1ea40e5d29534d3631bc3d15e78b/multibody/tree/rotational_inertia.h#L591)):*
> ```objc
>   /// calling CouldBePhysicallyValid() when the about-point is Bcm (the body's
>   /// center of mass).  Note: this class does not know its about-point or its
>   /// center of mass location.
> ```

BTW maybe note here that the mass_dcm_squared parameter is intended to provide that information (effectively)
___
*[`multibody/tree/test/spatial_inertia_test.cc` line 953 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQuGew3wcVbOFkS9gP:-NXQuGew3wcVbOFkS9gQ:b2vb74e) ([raw file](https://github.com/RobotLocomotion/drake/blob/b84ac3b9d5ea1ea40e5d29534d3631bc3d15e78b/multibody/tree/test/spatial_inertia_test.cc#L953)):*
> ```cpp
>   // I_BBcm_B calculates principal moments of inertia of
>   // [0.0  20.833333730697632  20.83333373069763] which violates the triangle
>   // inequality by ≈ 1.19E-06. Note the ratio 1.19E-06 / 1.14E-12 ≈ 10³ = 1000.
> ```

BTW Hmmm. This seems like such a huge violation I'm thinking we should reject it because of the potential downstream consequences. WDYT? Should we set a maximum limit on the error we'll tolerate? Maybe insist on 1e-10 or better? Your test above already has a 10 kg part shifted 300 meters!

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551)* status: 3 unresolved discussions, LGTM missing from assignee sherm1(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQwc928kBqCU2oXRvI:-NXQwc928kBqCU2oXRvJ:b-t9p0sp) (no related file):*
Naively, my expectation was that `MakeFromCentralInertia` could be improved to _maintain_ the triangle inequality (strictly, without any tolerancing), so that anyone who was _relying_ on the triangle inequality could continue to do so.  Here now, we have the possible hazard where downstream code that was assuming that the inequality held with no tolerance would be broken by this change.  Could you help me clarify my thinking around that, or point out where I went wrong?

___

**Review by mitiguy — COMMENTED** [review]
Hmmm -- still pondering whether we should implement an absolute cutoff.  

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-:-NXRYxew8iWFz1QzpVtH:bc95hie)* status: 2 unresolved discussions, LGTM missing from assignee sherm1(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQwc928kBqCU2oXRvI:-NXRJD6QAzoHwYCe1gKR:b-ebziml) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Naively, my expectation was that `MakeFromCentralInertia` could be improved to _maintain_ the triangle inequality (strictly, without any tolerancing), so that anyone who was _relying_ on the triangle inequality could continue to do so.  Here now, we have the possible hazard where downstream code that was assuming that the inequality held with no tolerance would be broken by this change.  Could you help me clarify my thinking around that, or point out where I went wrong?
</blockquote></details>

I think there is at least one more paths that can cause similar issues.  
I added another test in spatial_inertia_test.cc that may clarify similar issues with this constructor:
SpatialInertia(const T& mass, const Vector3<T>& p_PScm_E, const UnitInertia<T>& G_SP_E,
      const bool skip_validity_check = false).

___
___
*[`multibody/tree/rotational_inertia.h` line 591 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQsH870Um-DKl4XNZ8:-NXRXHZeBlmxdesoRppJ:bb6un1j) ([raw file](https://github.com/RobotLocomotion/drake/blob/b84ac3b9d5ea1ea40e5d29534d3631bc3d15e78b/multibody/tree/rotational_inertia.h#L591)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW maybe note here that the mass_dcm_squared parameter is intended to provide that information (effectively)
</blockquote></details>

Done

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551)* status: 2 unresolved discussions, LGTM missing from assignee sherm1(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQwc928kBqCU2oXRvI:-NXRZ_ySEEMQ8qtIs-0L:b5h3449) (no related file):*
<details><summary><i>Previously, mitiguy (Mitiguy) wrote…</i></summary><blockquote>

I think there is at least one more paths that can cause similar issues.  
I added another test in spatial_inertia_test.cc that may clarify similar issues with this constructor:
SpatialInertia(const T& mass, const Vector3<T>& p_PScm_E, const UnitInertia<T>& G_SP_E,
      const bool skip_validity_check = false).
</blockquote></details>

That's not responsive to my question though.  Having to fix a few functions to correct the numbers isn't any worse than fixing just that one place.

Perhaps an analogy would help: We have other math functions that take a unit vector as an argument.  If the argument is too far from the unit norm, we will throw.  However, if the argument is within some tolerance of the norm, then we'll normalize it and proceed.  My question here is for our inertia constructors and factory methods, so long as the arguments are within this small tolerance of the inequality criterion, why don't we automatically normalize the numbers and then proceed?  Why do we leave the invalid number intact and hand a newly-constructed object back to the user that violates its invariants?

___

**Review by sherm1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551)* status: 2 unresolved discussions, LGTM missing from assignee sherm1(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQwc928kBqCU2oXRvI:-NXRa4g899TFPtJuSh09:b-5u9vfd) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

That's not responsive to my question though.  Having to fix a few functions to correct the numbers isn't any worse than fixing just that one place.

Perhaps an analogy would help: We have other math functions that take a unit vector as an argument.  If the argument is too far from the unit norm, we will throw.  However, if the argument is within some tolerance of the norm, then we'll normalize it and proceed.  My question here is for our inertia constructors and factory methods, so long as the arguments are within this small tolerance of the inequality criterion, why don't we automatically normalize the numbers and then proceed?  Why do we leave the invalid number intact and hand a newly-constructed object back to the user that violates its invariants?
</blockquote></details>

I was also hoping for some kind of symmetry-preserving transformation. Maybe that could be done for some special cases? I'm not sure how though. What's happening here is that the inertia was provided as a beautiful central inertia with pattern (a, a, 0). We store that as an inertia about the body origin, requiring a large shift and producing a general inertia with no obvious pattern. Then when asked to do the check, we shift to CoM. Now the original pattern is damaged by roundoff caused by the two shifts. WDYT @mitiguy?

___

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551)* status: 2 unresolved discussions, LGTM missing from assignee sherm1(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQwc928kBqCU2oXRvI:-NXRgbdc-AiJMmuw0tWc:bpltqnm) (no related file):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I was also hoping for some kind of symmetry-preserving transformation. Maybe that could be done for some special cases? I'm not sure how though. What's happening here is that the inertia was provided as a beautiful central inertia with pattern (a, a, 0). We store that as an inertia about the body origin, requiring a large shift and producing a general inertia with no obvious pattern. Then when asked to do the check, we shift to CoM. Now the original pattern is damaged by roundoff caused by the two shifts. WDYT @mitiguy?
</blockquote></details>

I was imagining something like solving for the adjustments `d,e,f` that minimize `norm([d,e,f])` subject to "`[ixx+d,iyy+e,izz+f]` satisfies the triangle inequality" -- projecting the barely-invalid moments to the closest nearby moments.  (That would be a slightly more refined version of just "whatever the epsilon violation was, add it to one of the moments".)

In other words, maybe it would be OK if the second shift (back to the CoM) ended up with something within some epsilon of `[a, a, 0]` instead of exactly `[a, a, 0]`, as a tradeoff for the body-origin inertia being strictly physically valid.

___

**Review by sherm1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19551)* status: 2 unresolved discussions, LGTM missing from assignee sherm1(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @mitiguy)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19551#-NXQwc928kBqCU2oXRvI:-NXS7lDo27-wr-g119GD:b-572yjm) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I was imagining something like solving for the adjustments `d,e,f` that minimize `norm([d,e,f])` subject to "`[ixx+d,iyy+e,izz+f]` satisfies the triangle inequality" -- projecting the barely-invalid moments to the closest nearby moments.  (That would be a slightly more refined version of just "whatever the epsilon violation was, add it to one of the moments".)

In other words, maybe it would be OK if the second shift (back to the CoM) ended up with something within some epsilon of `[a, a, 0]` instead of exactly `[a, a, 0]`, as a tradeoff for the body-origin inertia being strictly physically valid.
</blockquote></details>

Paul and I discussed this and decided we should abandon this approach. It has a fatal flaw: imagine someone gives us a slightly bad central inertia, say (1e-8, 0, 0), but claims the COM is far from the body origin, say (10, 20, 100). The current algorithm will use the shift vector to loosen the tolerance check so will then declare the inertia valid. That's not good!

Instead, here is a better approach: if a user provides a central inertia, do the validity check immediately on the central inertia rather than by first shifting to the body origin and then shifting back to COM to do the check. While that may not solve all possible problems, it should be a very satisfactory solution to inertias that come from sdf and urdf (and most other places) where central inertias are the coin of the realm.

___
